pkg/e2etest/e2e_test.go

```diff
@@ -0,0 +1,529 @@
+			panic(err)
+		}
+		copyDone <- struct{}{}
+	}()
+
+	command := cmd.GetRootCmd()
+	command.SetArgs([]string{
+		"-i", strings.Join(mockIfaces.Names(), ","),
+		"-e", "json",
+		"-l", time.Now().Add(time.Hour).Format(time.ANSIC),
+		"-d", config.RuntimeDBPath(),
+		"-n", strconv.Itoa(maxEntries),
+		"-s", "packets",
+		"sip,dip,dport,proto",
+	})
+
+	require.Nil(t, command.Execute())
+	require.Nil(t, wr.Close())
+	<-copyDone
+	os.Stdout = old // restore the inital STDOUT
+
+	var res results.Result
+	require.Nil(t, jsoniter.NewDecoder(buf).Decode(&res))
+
+	return res
+}
+
+func setupSources(t testing.TB, ifaces mockIfaces) func() (mockIfaces, func(c *capture.Capture) (slimcap.Source, error)) {
+
+	fnMap := make(map[string]func(c *capture.Capture) (slimcap.Source, error))
+	for _, mockIface := range ifaces {
+		fnMap[mockIface.name] = mockIface.sourceInitFn
+	}
+
+	return func() (mockIfaces, func(c *capture.Capture) (slimcap.Source, error)) {
+		return ifaces, func(c *capture.Capture) (slimcap.Source, error) {
+			mockIfaceFn, ok := fnMap[c.Iface()]
+			if !ok {
+				return nil, fmt.Errorf("unable to find interface `%s` in list of mock interfaces", c.Iface())
+			}
+
+			return mockIfaceFn(c)
+		}
+	}
+}
+
+func newPcapSource(t testing.TB, name string, data []byte) (res *mockIface) {
+
+	res = &mockIface{
+		name: name,
+		src:  &afring.MockSource{},
+		tracking: &mockTracking{
+			done: make(chan struct{}, 1),
+		},
+		flows:   &map[capturetypes.EPHash]types.Counters{},
+		RWMutex: sync.RWMutex{},
+	}
+
+	res.sourceInitFn = func(c *capture.Capture) (slimcap.Source, error) {
+
+		res.Lock()
+		defer res.Unlock()
+
+		src, err := pcap.NewSource(res.name, bytes.NewBuffer(data))
+		require.Nil(t, err)
+		src.PacketAddCallbackFn(func(payload []byte, totalLen uint32, pktType, ipLayerOffset byte) {
+
+			res.Lock()
+			defer res.Unlock()
+
+			res.tracking.nRead++
+		})
+
+		mockSrc, err := afring.NewMockSource(c.Iface(),
+			afring.CaptureLength(link.CaptureLengthMinimalIPv6Transport),
+			afring.Promiscuous(false),
+			afring.BufferSize(1024*1024, 4),
+		)
+		require.Nil(t, err)
+		*res.src = *mockSrc
+
+		pkt := mockSrc.NewPacket()
+		mockSrc.PacketAddCallbackFn(func(payload []byte, totalLen uint32, pktType, ipLayerOffset byte) {
+
+			res.Lock()
+			defer res.Unlock()
+
+			pkt = slimcap.NewIPPacket(pkt, payload, pktType, int(totalLen), ipLayerOffset)
+			var p capturetypes.GPPacket
+			if err := capture.Populate(&p, pkt); err != nil {
+				res.tracking.nErr++
+				return
+			}
+
+			hash, hashReverse := p.EPHash, p.EPHashReverse
+			if direction := capturetypes.ClassifyPacketDirection(&p); direction != capturetypes.DirectionUnknown {
+				if direction == capturetypes.DirectionReverts || direction == capturetypes.DirectionMaybeReverts {
+					hash, hashReverse = p.EPHashReverse, p.EPHash
+				}
+			}
+
+			hash[34], hash[35] = 0, 0
+			hashReverse[34], hashReverse[35] = 0, 0
+
+			if flow, exists := (*res.flows)[hash]; exists {
```

Comment:
This can be cleaned up a bit I guess, but I _deliberately_ re-implemented a crude flow logic to avoid using too much of the core code in the E2E test (otherwise the chance is high that we miss out on mistakes because we auto-correct it).